### ISSUE TITLE
Non overlapping PRs

### ISSUE THREAD

**Issue by spence-novata** [facebook/sapling#302]
Would people be favorable to a PR strategy where each PR's base is the head of the previous PR?

So instead of 

```
pr1 -> master
pr2 -> master
pr3 -> master
pr4 -> master
```

You have

pr1 -> master
pr2 -> pr1
pr3 -> p2
pr4 -> p3

This presents a few benefits over the current strategy of overlapping PRs:

* PRs can be merged on GitHub (you still wouldn't want to merge out of order, and [draft status PRs](https://github.com/facebook/sapling/issues/288) would help manage this)
* As you merge previous PRs and delete their branches, GitHub automatically updates the base of the next PR to the branch the previous PR was merged into (so after you merge pr1, GitHub will magically make your PRs pr2 -> master, pr3 -> pr2, pr4 -> p3)
* The changes of each commit can be reviewed in isolation on GitHub. The work on ReviewStack is admirable, but it doesn't feel as mature as the Sapling CLI and hasn't reached parity with GitHub PRs regarding features or experience.

**Comment by spence-novata** [comment]
Is it not possible to offer a non-overlapping strategy behind configuration for those that do have the required access?

**Comment by spence-novata** [comment]
Yeah the whole base/head branch thing that ghstack does doesn't really leverage the way GitHub automatically changes the base as branches are merged and deleted and still means you can't merge in the GitHub UI

I'm happy to force push as this is what I already do manually.

**Comment by bolinfest** [comment]
Honestly, I think I would prefer this, as well, particularly for personal projects. Part of the reason force-pushing causes problems in GitHub today is that GitHub periodically reaps commits that are not part of any named branch, so force-pushing "abandons" your old branch. I added an almost comical workaround for this in `sl pr`:

https://github.com/facebook/sapling/issues/302

Though even if we support this new mode of stacked pull requests, I tend to think `sl pr` still needs to be default because it's the only thing that works *everywhere* because it does not require write access.

**Comment by zengk95** [comment]
+1 to this. I think I suggested it in here https://github.com/facebook/sapling/issues/284 

My use case is on a GHE instance and review stack is not supported. This makes stacked PRs hard to review cause PRs later on just contain all the changes prior.

**Comment by bolinfest** [comment]
@spence-novata @zengk95 as noted in 166e2640d3, I have experimental support for this workflow. If you're willing to build from source, you can try it today, or you can wait for the next release? I would love to get feedback on this approach!

**Comment by FirelightFlagboy** [comment]
Hello, does this experimental workflow still available on `0.2.20231113` ? 

I would like to try it but the configuration `github.pr_workflow` isn't listed in `sl config --verbose` and the workflow provided by `ghstack` isn't something I want.

Regards

**Comment by oleg-codaio** [comment]
This works now (running 0.2.20240219); run: `sl config --local github.pr_workflow single`.

However, merging the bottom-most PR in the stack results in a merge conflict on the next PR:

<img width="842" alt="image" src="https://github.com/facebook/sapling/assets/1815707/d949fb38-1122-4ff3-a0da-0e88b17bc8b1">

Even if it didn't, the next PR still has commits from downstack PRs. To fully support this workflow there might be two solutions:

1. When merging a stack at a given PR, change the base of that PR to `main` and merge that, then close all the PRs below it in the stack. This is the approach that https://github.com/ejoffe/spr takes with `git spr merge`.
2. Alternatively, add some `sl pr merge` command that starts from the bottom of the stack, merges the PR, then rebases the stack, then repeats for all the remaining PRs in the stack (waiting for status checks as needed). With a local-only approach this is bottlenecked on CI; Graphite has done something similar but that process happens on their end so it's much more convenient

**Comment by shikanime** [comment]
One thing I've been doing is setting the [close keyword](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) just before merging a pull request to close bottom Pull Requests. Could adding this keyword to each child PR be a possible solution for overlap mode?

**Comment by shikanime** [comment]
I just noticed that when using overlap PR with histedit marks them as closes or merged
